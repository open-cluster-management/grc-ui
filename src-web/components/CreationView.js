/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ******************************************************************************
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { hideResourceToolbar } from '../../lib/client/resource-helper'
import { TemplateEditor } from './TemplateEditor'
import policyTemplate from './templates/policy-template.hbs'
import Choices from './templates'
import msgs from '../../nls/platform.properties'
import _ from 'lodash'

// where to put Create/Cancel buttons
const Portals = Object.freeze({
  editBtn: 'edit-button-portal-id',
  createBtn: 'create-button-portal-id',
  cancelBtn: 'cancel-button-portal-id',
})

// controlData is converted to templateData when tempplate handlebar is rendered
//  id: becomes template variable;
//  active: becomes value replacement
//  type: what type of control to prompt user
//  available: available choices
//  reverse: the yaml object to control path--if user edits yaml, what control should be updated
//  mustExist: when validating the template for creation, this value must exist
const controlData = [
  {
    name: 'creation.view.policy.name',
    id: 'name',
    type: 'text',
    active: 'policy-grc',
    reverse: 'Policy[0].metadata.name',
    mustExist: true,
  },
  {
    name: 'creation.view.policy.namespace',
    id: 'namespace',
    type: 'singleselect',
    description: 'policy.create.namespace.tooltip',
    available: [],
    reverse: 'Policy[0].metadata.namespace',
    mustExist: true,
  },
  {
    name: 'creation.view.policy.specs',
    description: 'policy.create.specs.tooltip',
    placeholder: 'creation.view.policy.select.specs',
    id: 'specs',
    type: 'multiselect',
    available: [],
    isOneSelection: true, // close dropdown on one selection--otherwise dropdown stays open
    updateNamePrefix: 'policy-', // if user hasn't typed in a custom name, update name using this selection
    reverse: ['Policy[0].spec',],
    mustExist: true,
  },
  {
    name: 'creation.view.policy.binding',
    description: 'policy.create.selectors.tooltip',
    placeholder: 'creation.view.policy.select.selectors',
    id: 'clusters',
    type: 'multiselect',
    available: [],
    reverse: 'PlacementRule[0].spec.clusterSelector', // automatically does matchLabels && matchExpressions
    mustExist: false,
  },
  {
    name: 'creation.view.policy.standards',
    description: 'policy.create.standards.tooltip',
    placeholder: 'creation.view.policy.select.standards',
    id: 'standards',
    type: 'multiselect',
    available: ['NIST', 'NIST-CSF', 'PCI', 'FISMA', 'HIPAA'],
    reverse: 'Policy[0].metadata.annotations["policy.open-cluster-management.io/standards"]',
    cacheUserValueKey: 'create.policy.standards',
  },
  {
    name: 'creation.view.policy.categories',
    description: 'policy.create.categories.tooltip',
    placeholder: 'creation.view.policy.select.categories',
    id: 'categories',
    type: 'multiselect',
    available: ['PR.PT Protective Technology','PR.DS Data Security', 'PR.AC Identity Management Authentication and Access Control', 'PR.IP Information Protection Processes and Procedures', 'DE.CM Security Continuous Monitoring'],
    reverse: 'Policy[0].metadata.annotations["policy.open-cluster-management.io/categories"]',
    cacheUserValueKey: 'create.policy.categories',
  },
  {
    name: 'creation.view.policy.controls',
    description: 'policy.create.controls.tooltip',
    placeholder: 'creation.view.policy.select.controls',
    id: 'controls',
    type: 'multiselect',
    available: ['PR.PT-1 Audit Logging','PR.PT-3 Least Functionality','PR.DS-2 Data-in-transit','PR.DS-2 Data-at-rest','PR.AC-4 Access Control', 'PR.AC-5 Network Integrity', 'PR.IP-1 Baseline Configuration', 'DE.CM-7 Monitoring for unauthorized activity','DE.CM-8 Vulnerability scans'],
    reverse: 'Policy[0].metadata.annotations["policy.open-cluster-management.io/controls"]',
    cacheUserValueKey: 'create.policy.controls',
  },
  {
    name: 'creation.view.policy.enforce',
    description: 'policy.create.enforce.tooltip',
    id: 'enforce',
    type: 'checkbox',
    active: 'inform',
    available: ['inform', 'enforce'],  // in template, 'inform'===checkbox unchecked
    reverse: 'Policy[0].spec.remediationAction',
    mustExist: true,
  },
  {
    name: 'creation.view.policy.disabled',
    description: 'policy.create.disabled.tooltip',
    id: 'disabled',
    type: 'checkbox',
    active: 'false',
    available: ['false', 'true'],
    reverse: 'Policy[0].spec.disabled',
  },
]


export default class CreationView extends React.Component {

  static propTypes = {
    createControl: PropTypes.shape({
      createResource: PropTypes.func,
      cancelCreate: PropTypes.func,
      creationStatus: PropTypes.string,
      creationMsg: PropTypes.string
    }),
    discovered: PropTypes.object,
    fetchControl: PropTypes.shape({
      isLoaded: PropTypes.bool,
      isFailed: PropTypes.bool,
    }),
  }

  render() {
    hideResourceToolbar()
    const { locale } = this.context
    const {fetchControl, createControl, discovered} = this.props
    return (
      <TemplateEditor
        template={policyTemplate}
        controlData={getControlData(discovered, locale)}
        portals={Portals}
        fetchControl={fetchControl}
        createControl={createControl}
        type={'policy'}
        locale={locale}
      />
    )
  }
}

const getControlData = (discovered, locale) => {
  if (discovered) {
    const mergedData = _.cloneDeep(controlData)

    // add preset spec choices from yaml
    const controlMap = _.keyBy(mergedData, 'id')
    Object.values(Choices).forEach(choice=>{
      const available = _.get(controlMap, `${choice.multiselect}.available`)
      if (available) {
        available.push(choice)
      }
    })

    // add discovered choices from server
    //  add available annotations to categories, etc controls
    //  add existing policy names to name control
    const {policyNames, namespaces, annotations, clusterLabels} = discovered
    const {name, namespace, clusters, standards, categories, controls } = _.keyBy(mergedData, 'id')
    name.existing = policyNames
    namespace.available = namespaces
    clusters.available = clusterLabels
    standards.available = [...new Set([...standards.available, ...annotations.standards])]
    categories.available = [...new Set([...categories.available, ...annotations.categories])]
    controls.available = [...new Set([...controls.available, ...annotations.controls])]

    // convert message keys
    mergedData.forEach(control=>{
      ['name', 'description', 'placeholder'].forEach(key=>{
        if (control[key]) {
          control[key] = msgs.get(control[key], locale)
        }
      })
    })

    return mergedData
  }
  return controlData
}
