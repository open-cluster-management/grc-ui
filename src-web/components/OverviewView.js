/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************
 /* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { updateResourceToolbar, updateActiveFilters } from '../actions/common'
import { Loading, Notification } from 'carbon-components-react'
import { filterPolicies, filterFindings, getAvailableGrcFilters, getSavedGrcState, saveGrcState, combineResourceFilters,replaceGrcState } from '../../lib/client/filter-helper'
import { showResourceToolbar, hideResourceToolbar } from '../../lib/client/resource-helper'
import { GRC_VIEW_STATE_COOKIE, GRC_FILTER_STATE_COOKIE } from '../../lib/shared/constants'
import RecentActivityModule from './modules/RecentActivityModule'
import RecentActivityModuleWithoutFindings from './modules/RecentActivityModuleWithoutFindings'
import ImpactedControlsModule from './modules/ImpactedControlsModule'
import PolicySummaryModule from './modules/PolicySummaryModule'
import config from '../../lib/shared/config'
import msgs from '../../nls/platform.properties'
import _ from 'lodash'
import NoResource from '../components/common/NoResource'
import ResourceFilterBar from '../components/common/ResourceFilterBar'
import createDocLink from '../components/common/CreateDocLink'
import queryString from 'query-string'

export class OverviewView extends React.Component {

  constructor (props) {
    super(props)
    this.state= {
      viewState: getSavedGrcState(GRC_VIEW_STATE_COOKIE)
    }
    this.scroll = _.debounce(()=>{
      this.scrollView()
    }, 50)
    this.onUnload = this.onUnload.bind(this)
    window.addEventListener('beforeunload', this.onUnload)
    this.updateViewState = this.updateViewState.bind(this)
    this.handleCreatePolicy = this.handleCreatePolicy.bind(this)
    this.handleDrillDownClickOverview = this.handleDrillDownClickOverview.bind(this)
  }

  componentWillMount() {
    const { activeFilters={} } = this.props
    //get (activeFilters ∪ storedFilters) only since availableGrcFilters is uninitialized at this stage
    //later when availableGrcFilters initialized, will do further filtering in componentWillReceiveProps
    const combinedFilters = combineResourceFilters(activeFilters, getSavedGrcState(GRC_FILTER_STATE_COOKIE))
    delete combinedFilters.severity//remove severity filter set during first time render
    //update sessionStorage
    replaceGrcState(GRC_FILTER_STATE_COOKIE, combinedFilters)
    //update active filters
    updateActiveFilters(combinedFilters)
  }

  componentDidMount() {
    window.addEventListener('scroll', this.scroll)
  }

  componentWillUnmount() {
    window.removeEventListener('beforeunload', this.onUnload)
    window.removeEventListener('scroll', this.scroll)
    this.onUnload()
  }

  componentWillReceiveProps(nextProps) {
    const {refreshControl, policies, findings, updateActiveFilters, updateResourceToolbar} = nextProps
    if (!_.isEqual(refreshControl, this.props.refreshControl) ||
        !_.isEqual(policies, this.props.policies)) {
      const { locale } = this.context
      const availableGrcFilters = getAvailableGrcFilters(policies, findings, locale)
      updateResourceToolbar(refreshControl, availableGrcFilters)
      const activeFilters = _.cloneDeep(nextProps.activeFilters||{})
      //get (activeFilters ∪ storedFilters) ∩ availableGrcFilters
      const combinedFilters = combineResourceFilters(activeFilters, getSavedGrcState(GRC_FILTER_STATE_COOKIE), availableGrcFilters)
      //update sessionStorage
      replaceGrcState(GRC_FILTER_STATE_COOKIE, combinedFilters)
      //update active filters
      updateActiveFilters(combinedFilters)
    }
  }

  render() {
    const { locale } = this.context
    const showFindings = config['feature_security-findings']
    const { loading, error, policies, findings, applications, activeFilters={} } = this.props
    hideResourceToolbar()

    if (loading) {
      return <Loading withOverlay={false} className='content-spinner' />
    }

    if (error) {
      if (error.name === 'PermissionError') {
        return <Notification title='' className='overview-notification' kind='error'
          subtitle={msgs.get('error.permission.denied', locale)} />
      }
      return <Notification title='' className='overview-notification' kind='error'
        subtitle={msgs.get('overview.error.default', locale)} />
    }

    if ((!policies || policies.length === 0) && !loading) {
      return (
        <NoResource
          title={msgs.get('no-resource.title', [msgs.get('routes.grc', locale)], locale)}
          detail={msgs.get('no-resource.detail.policy', locale)}>
          {createDocLink(locale, this.handleCreatePolicy, msgs.get('routes.create.policy', locale))}
        </NoResource>
      )
    }
    showResourceToolbar()
    const { viewState } = this.state
    const showApplications = this.props.showApplications
    const availableFilters =  getAvailableGrcFilters(policies, findings, locale)
    const filteredPolicies = filterPolicies(policies, activeFilters, locale, 'metadata.annotations')
    const filteredFindings = filterFindings(findings, activeFilters, locale)
    //need to filteredapplications later?
    if(showFindings === false) {
      return (
        <div className='overview-view'>
          <ResourceFilterBar />
          <RecentActivityModuleWithoutFindings
            showApplications={showApplications}
            policies={filteredPolicies}
            showFindings={showFindings}
            findings={filteredFindings}
            applications={applications}
            handleDrillDownClick={this.handleDrillDownClickOverview}
            viewState={viewState}
            updateViewState={this.updateViewState} />
          <ImpactedControlsModule
            viewState={viewState}
            updateViewState={this.updateViewState}
            policies={filteredPolicies}
            showFindings={showFindings}
            findings={filteredFindings}
            activeFilters={activeFilters}
            availableFilters={availableFilters}
            handleDrillDownClick={this.handleDrillDownClickOverview} />
          <PolicySummaryModule
            policies={filteredPolicies}
            findings={filteredFindings} />
        </div>
      )
    }
    else {
      return (
        <div className='overview-view'>
          <ResourceFilterBar />
          <RecentActivityModule
            showApplications={showApplications}
            policies={filteredPolicies}
            showFindings={showFindings}
            findings={filteredFindings}
            applications={applications}
            handleDrillDownClick={this.handleDrillDownClickOverview}
            viewState={viewState}
            updateViewState={this.updateViewState} />
          <ImpactedControlsModule
            viewState={viewState}
            updateViewState={this.updateViewState}
            policies={filteredPolicies}
            showFindings={showFindings}
            findings={filteredFindings}
            activeFilters={activeFilters}
            availableFilters={availableFilters}
            handleDrillDownClick={this.handleDrillDownClickOverview} />
          <PolicySummaryModule
            policies={filteredPolicies}
            findings={filteredFindings} />
        </div>
      )
    }
  }

  updateViewState(states) {
    this.setState(prevState=>{
      return {viewState:  Object.assign(prevState.viewState, states)}
    })
  }

  onUnload() {//saved overview ui setting
    saveGrcState(GRC_VIEW_STATE_COOKIE, this.state.viewState)
  }

  scrollView () {
    const headerRef = document.getElementsByClassName('secondary-header')[0]
    const contentRef = document.getElementsByClassName('overview-view')[0]
    if (headerRef && contentRef) {
      const contentRect = contentRef.getBoundingClientRect()
      headerRef.classList.toggle('bottom-border', contentRect.top<180)
    }
  }

  handleCreatePolicy(){
    this.props.history.push(`${config.contextPath}/create`)
  }

  handleDrillDownClickOverview(parentType, parentName){
    if(parentType){
      let page = 'all'
      const paraURL = {}
      switch(parentType.toLowerCase()){
      case 'policy violations'://RecentActivityModule to policies page, all policy violations
        paraURL.index = 0
        paraURL.autoFocus = 'module-toggle-tab'
        break
      case 'cluster violations'://RecentActivityModule to policies page, all cluster violations
        paraURL.index = 1
        paraURL.autoFocus = 'module-toggle-tab'
        break
      case 'high'://RecentActivityModule to security findings page, all high severity findings
      case 'medium'://all medium severity findings
      case 'low'://all low severity findings
        page = 'findings'
        paraURL.index = 0
        paraURL.autoFocus = 'module-toggle-tab'
        paraURL.severity = _.startCase(parentType.toLowerCase())
        break
      case 'policies'://TopInformationModule to policies page with specific policy violation name
        paraURL.index = 0
        paraURL.autoFocus = 'module-toggle-tab'
        if(parentName){
          // paraURL.name = parentName //highlight this policy violation name
          // paraURL.side = true //auto open side panel under this highlight name
          paraURL.filters = `{"textsearch":["${parentName}"]}`
        }
        break
      case 'clusters'://TopInformationModule to policies page with specific cluster violation name
        paraURL.index = 1
        paraURL.autoFocus = 'module-toggle-tab'
        if(parentName){
          // paraURL.name = parentName //highlight this cluster violation name
          // paraURL.side = true //auto open side panel under this highlight name
          paraURL.filters = `{"textsearch":["${parentName}"]}`
        }
        break
      case 'applications':
        paraURL.index = 2
        paraURL.autoFocus = 'module-toggle-tab'
        if(parentName){
          paraURL.filters = `{"textsearch":["${parentName}"]}`
        }
        break
      case 'security findings'://TopInformationModule to security findings page with specific security finding name
        page = 'findings'
        paraURL.index = 0
        paraURL.autoFocus = 'module-toggle-tab'
        if(parentName){
          paraURL.filters = `{"textsearch":["${parentName}"]}` //current no highlight and side panel for security findings page, just filter name
        }
        break
      case 'finding clusters'://TopInformationModule to security findings page with specific finding cluster name
        page = 'findings'
        paraURL.index = 1
        paraURL.autoFocus = 'module-toggle-tab'
        if(parentName){
          paraURL.filters = `{"textsearch":["${parentName}"]}` //current no highlight and side panel for finding clusters page, just filter name
        }
        break
      default:
        break
      }
      const slash = this.props.location.pathname.trim().substr(-1) === '/' ? '' : '/'
      if (this.props.history) {
        this.props.history.push(`${this.props.location.pathname}${slash}${page}?${queryString.stringify(paraURL)}`)
      }
      return `${slash}${page}?${queryString.stringify(paraURL)}`
    }
  }
}

OverviewView.propTypes = {
  activeFilters: PropTypes.object,
  applications: PropTypes.array,
  error: PropTypes.object,
  findings: PropTypes.array,
  history: PropTypes.object.isRequired,
  loading: PropTypes.bool,
  location: PropTypes.object,
  policies: PropTypes.array,
  refreshControl: PropTypes.object,
  showApplications: PropTypes.bool,
  updateActiveFilters: PropTypes.func,
  updateResourceToolbar: PropTypes.func,
}

const mapStateToProps = (state) => {
  const {resourceToolbar: {activeFilters}} = state
  return { activeFilters }
}

const mapDispatchToProps = (dispatch) => {
  return {
    updateResourceToolbar: (refreshControl, availableFilters) => dispatch(updateResourceToolbar(refreshControl, availableFilters)),
    updateActiveFilters: (activeFilters) => dispatch(updateActiveFilters(activeFilters))
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(OverviewView))
