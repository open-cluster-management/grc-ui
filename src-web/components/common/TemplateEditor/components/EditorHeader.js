/*******************************************************************************
 * Licensed Materials - Property of IBM
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import '../scss/editor-header.scss'
import msgs from '../../../../nls/platform.properties'

class EditorHeader extends React.Component {

  static propTypes = {
    children: PropTypes.node,
    locale: PropTypes.string,
  }

  constructor (props) {
    super(props)
  }

  render() {
    const { children, locale } = this.props
    const editorToolbarTitle = msgs.get('editor.toolbar', locale)
    return (
      <div className='creation-view-yaml-header' >
        <div className='creation-view-yaml-header-toolbar' role='region' aria-label={editorToolbarTitle} id={editorToolbarTitle}>
          {children}
        </div>
      </div>
    )
  }

}

export default EditorHeader
