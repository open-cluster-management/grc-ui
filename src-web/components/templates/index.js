/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import specPodRole from './spec-rolebinding.yaml'
import specPodSecurity from './spec-psp.yaml'
import specPodNetwork from './spec-network.yaml'
import specPodLimit from './spec-mem-limit.yaml'
import specPodExists from './spec-pod.yaml'
import specRoles from './spec-roles.yaml'
import specNamespace from './spec-namespace.yaml'
import specClustadminrole from './spec-clusteradminrole.yaml'
import specCertmgmtexp from './spec-certmgmtexp.yaml'
import specMutationAdvisor from './spec-mutationadvisor.yaml'
import specVAsecurity from './spec-vasecurity.yaml'
import specKubeEncryption from './spec-kubeencryption.yaml'
import specAudit from './spec-audit.yaml'
import specCis from './spec-cis.yaml'
import specCisOcp from './spec-cis-ocp.yaml'

export {
  specPodRole,
  specPodSecurity,
  specPodNetwork,
  specPodLimit,
  specPodExists,
  specRoles,
  specNamespace,
  specClustadminrole,
  specCertmgmtexp,
  specMutationAdvisor,
  specVAsecurity,
  specKubeEncryption,
  specAudit,
  specCis,
  specCisOcp,
}
