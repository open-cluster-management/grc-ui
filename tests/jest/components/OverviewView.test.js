/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { GrcViewPolicyCluster, GrcViewFindingCluster, GrcViewRefreshControl, GrcViewSecondaryHeaderProps } from './ComponentsTestingData'
//curly braces means pure component without redux
//which is what we want in unit test
import { OverviewView } from '../../../src-web/components/OverviewView'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'
import { shallow } from 'enzyme'
import { createMemoryHistory } from 'history'

const history = createMemoryHistory({'length':4,'action':'POP','location':{'pathname':'/multicloud/policies/','search':'','hash':''}})

describe('OverviewView component 1', () => {
  const location = {
    pathname: '/multicloud/policies/'
  }
  it('renders as expected', () => {
    const component = renderer.create(
      <BrowserRouter>
        <OverviewView
          title='Test1'
          history={history}
          location={location}
          loading={true}
          error={null}
          activeFilters={{}}
          policies={GrcViewPolicyCluster}
          findings={GrcViewFindingCluster}
          refreshControl={GrcViewRefreshControl}
          secondaryHeaderProps={GrcViewSecondaryHeaderProps}
        />
      </BrowserRouter>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('OverviewView component 2', () => {
  const location = {
    pathname: '/multicloud/policies/'
  }
  it('renders as expected', () => {
    const component = renderer.create(
      <BrowserRouter>
        <OverviewView
          title='Test2'
          history={history}
          location={location}
          loading={false}
          error={{}}
          activeFilters={{}}
          policies={GrcViewPolicyCluster}
          findings={GrcViewFindingCluster}
          refreshControl={GrcViewRefreshControl}
          secondaryHeaderProps={GrcViewSecondaryHeaderProps}
        />
      </BrowserRouter>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('OverviewView component 3', () => {
  const location = {
    pathname: '/multicloud/policies/'
  }
  it('renders as expected', () => {
    const component = renderer.create(
      <BrowserRouter>
        <OverviewView
          title='Test3'
          history={history}
          location={location}
          loading={false}
          error={null}
          activeFilters={{}}
          policies={[]}
          findings={GrcViewFindingCluster}
          refreshControl={GrcViewRefreshControl}
          secondaryHeaderProps={GrcViewSecondaryHeaderProps}
        />
      </BrowserRouter>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('OverviewView handleDrillDownClickOverview 1', () => {
  const location1 = {
    pathname: '/multicloud/policies/'
  }
  const location2 = {
    pathname: '/multicloud/policies'
  }
  it('renders as expected', () => {
    const wrapper = shallow(
      <OverviewView
        title='Test4'
        history={history}
        location={location1}
        loading={false}
        error={{}}
        activeFilters={{}}
        policies={GrcViewPolicyCluster}
        findings={GrcViewFindingCluster}
        refreshControl={GrcViewRefreshControl}
        secondaryHeaderProps={GrcViewSecondaryHeaderProps}
      />
    )
    expect(wrapper.instance().handleDrillDownClickOverview()).toEqual(undefined)
    expect(wrapper.instance().handleDrillDownClickOverview(null, 'cluster1')).toEqual(undefined)
    expect(wrapper.instance().handleDrillDownClickOverview('clusters')).toEqual('all?autoFocus=module-toggle-tab&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('clusters', 'cluster1')).toEqual('all?autoFocus=module-toggle-tab&filters=%7B%22textsearch%22%3A%5B%22cluster1%22%5D%7D&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('policies')).toEqual('all?autoFocus=module-toggle-tab&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('policies', 'policy-iampolicy')).toEqual('all?autoFocus=module-toggle-tab&filters=%7B%22textsearch%22%3A%5B%22policy-iampolicy%22%5D%7D&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('security findings')).toEqual('findings?autoFocus=module-toggle-tab&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('security findings', 'Policy that is not compliant')).toEqual('findings?autoFocus=module-toggle-tab&filters=%7B%22textsearch%22%3A%5B%22Policy%20that%20is%20not%20compliant%22%5D%7D&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('finding clusters')).toEqual('findings?autoFocus=module-toggle-tab&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('finding clusters', 'cluster1')).toEqual('findings?autoFocus=module-toggle-tab&filters=%7B%22textsearch%22%3A%5B%22cluster1%22%5D%7D&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('Policy Violations')).toEqual('all?autoFocus=module-toggle-tab&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('Cluster Violations')).toEqual('all?autoFocus=module-toggle-tab&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('High')).toEqual('findings?autoFocus=module-toggle-tab&index=0&severity=High')
    expect(wrapper.instance().handleDrillDownClickOverview('Medium')).toEqual('findings?autoFocus=module-toggle-tab&index=0&severity=Medium')
    expect(wrapper.instance().handleDrillDownClickOverview('Low')).toEqual('findings?autoFocus=module-toggle-tab&index=0&severity=Low')
  })
  it('renders as expected', () => {
    const wrapper = shallow(
      <OverviewView
        title='Test4'
        history={history}
        location={location2}
        loading={false}
        error={{}}
        activeFilters={{}}
        policies={GrcViewPolicyCluster}
        findings={GrcViewFindingCluster}
        refreshControl={GrcViewRefreshControl}
        secondaryHeaderProps={GrcViewSecondaryHeaderProps}
      />
    )
    expect(wrapper.instance().handleDrillDownClickOverview()).toEqual(undefined)
    expect(wrapper.instance().handleDrillDownClickOverview(null, 'cluster1')).toEqual(undefined)
    expect(wrapper.instance().handleDrillDownClickOverview('clusters')).toEqual('/all?autoFocus=module-toggle-tab&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('clusters', 'cluster1')).toEqual('/all?autoFocus=module-toggle-tab&filters=%7B%22textsearch%22%3A%5B%22cluster1%22%5D%7D&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('policies')).toEqual('/all?autoFocus=module-toggle-tab&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('policies', 'policy-iampolicy')).toEqual('/all?autoFocus=module-toggle-tab&filters=%7B%22textsearch%22%3A%5B%22policy-iampolicy%22%5D%7D&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('security findings')).toEqual('/findings?autoFocus=module-toggle-tab&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('security findings', 'Policy that is not compliant')).toEqual('/findings?autoFocus=module-toggle-tab&filters=%7B%22textsearch%22%3A%5B%22Policy%20that%20is%20not%20compliant%22%5D%7D&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('finding clusters')).toEqual('/findings?autoFocus=module-toggle-tab&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('finding clusters', 'cluster1')).toEqual('/findings?autoFocus=module-toggle-tab&filters=%7B%22textsearch%22%3A%5B%22cluster1%22%5D%7D&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('Policy Violations')).toEqual('/all?autoFocus=module-toggle-tab&index=0')
    expect(wrapper.instance().handleDrillDownClickOverview('Cluster Violations')).toEqual('/all?autoFocus=module-toggle-tab&index=1')
    expect(wrapper.instance().handleDrillDownClickOverview('High')).toEqual('/findings?autoFocus=module-toggle-tab&index=0&severity=High')
    expect(wrapper.instance().handleDrillDownClickOverview('Medium')).toEqual('/findings?autoFocus=module-toggle-tab&index=0&severity=Medium')
    expect(wrapper.instance().handleDrillDownClickOverview('Low')).toEqual('/findings?autoFocus=module-toggle-tab&index=0&severity=Low')
  })
})
