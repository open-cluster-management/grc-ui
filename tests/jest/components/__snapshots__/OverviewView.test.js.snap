// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewView component 1 renders as expected 1`] = `
<div
  aria-live="assertive"
  className="bx--loading content-spinner"
>
  <svg
    className="bx--loading__svg"
    viewBox="-75 -75 150 150"
  >
    <title>
      Loading
    </title>
    <circle
      className={null}
      cx="0"
      cy="0"
      r="37.5"
    />
  </svg>
</div>
`;

exports[`OverviewView component 2 renders as expected 1`] = `
<div
  className="bx--inline-notification bx--inline-notification--error overview-notification"
  kind="error"
  role="alert"
>
  <div
    className="bx--inline-notification__details"
  >
    <svg
      alt="closes notification"
      aria-label="closes notification"
      className="bx--inline-notification__icon"
      fill={undefined}
      fillRule="evenodd"
      height="16"
      name={undefined}
      role="img"
      style={undefined}
      viewBox="0 0 16 16"
      width="16"
    >
      <title>
        closes notification
      </title>
      <path
        d="M8 16A8 8 0 1 1 8 0a8 8 0 0 1 0 16zM3.293 4.707l8 8 1.414-1.414-8-8-1.414 1.414z"
      />
    </svg>
    <div
      className="bx--inline-notification__text-wrapper"
    >
      <p
        className="bx--inline-notification__title"
      >
        
      </p>
      <div
        className="bx--inline-notification__subtitle"
      >
        An unexpected error occurred. Try again.
      </div>
    </div>
  </div>
  <button
    className="bx--inline-notification__close-button"
    onClick={[Function]}
    title="close icon"
    type="button"
  >
    <svg
      alt="close icon"
      aria-label="close icon"
      className="bx--inline-notification__close-icon"
      fill={undefined}
      fillRule="evenodd"
      height="10"
      name={undefined}
      role="img"
      style={undefined}
      viewBox="0 0 10 10"
      width="10"
    >
      <title>
        close icon
      </title>
      <path
        d="M6.32 5L10 8.68 8.68 10 5 6.32 1.32 10 0 8.68 3.68 5 0 1.32 1.32 0 5 3.68 8.68 0 10 1.32 6.32 5z"
      />
    </svg>
  </button>
</div>
`;

exports[`OverviewView component 3 renders as expected 1`] = `
<div>
  <div
    className="no-resource"
  >
    <img
      alt="No resource"
      className="no-resource-icon"
      src="undefined/graphics/no-policy.svg"
    />
    <div
      className="no-resource-title"
    >
      Get set-up to track your governance and risk.
    </div>
    <div
      className="no-resource-detail"
    >
      We noticed that you didn't have any policies created yet. You can view our documentation to see some samples for creating new policies, or you can go ahead and create one now!
    </div>
    <div
      className="button-group"
    >
      <button
        className="bx--btn bx--btn--sm bx--btn--primary"
        disabled={false}
        id="create-resource"
        onClick={[Function]}
        tabIndex={0}
        type="button"
      >
        Create policy
        <svg
          alt="Provide icon description if icon is used"
          aria-hidden={true}
          aria-label="Provide icon description if icon is used"
          className="bx--btn__icon"
          fill={undefined}
          fillRule="evenodd"
          height="16"
          name="add--glyph"
          role="img"
          style={undefined}
          viewBox="0 0 16 16"
          width="16"
        >
          <title>
            Provide icon description if icon is used
          </title>
          <path
            d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm4 9H9v3H7V9H4V7h3V4h2v3h3v2z"
          />
        </svg>
      </button>
      <div
        className="buttons"
      >
        <button
          className="bx--btn bx--btn--sm bx--btn--secondary"
        >
          <a
            aria-describedby="launchWindow"
            className="doc-link"
            href="https://www-03preprod.ibm.com/support/knowledgecenter/SSBS6K_undefined/mcm/compliance/compliance_intro.html"
            target="doc"
          >
            View documentation
          </a>
        </button>
      </div>
    </div>
  </div>
</div>
`;
